src/v/cloud_storage/partition_manifest.cc

```diff
@@ -1164,26 +1168,68 @@ void partition_manifest::update(partition_manifest_handler&& handler) {
     }
 }
 
-serialized_json_stream partition_manifest::serialize() const {
+struct partition_manifest::serialization_cursor {
+    serialization_cursor(std::ostream& out, size_t max_segments)
+      : wrapper(out)
+      , writer(wrapper)
+      , max_segments_per_call(max_segments) {}
+
+    json::OStreamWrapper wrapper;
+    json::Writer<json::OStreamWrapper> writer;
+    // Next serialized offset
+    model::offset next_offset{};
+    size_t segments_serialized{0};
+    size_t max_segments_per_call{0};
+    // Flags that indicate serialization progress
+    bool prologue_done{false};
+    bool segments_done{false};
+    bool replaced_done{false};
+    bool epilogue_done{false};
+};
+
+ss::future<serialized_json_stream> partition_manifest::serialize() const {
     iobuf serialized;
     iobuf_ostreambuf obuf(serialized);
     std::ostream os(&obuf);
-    serialize(os);
+    serialization_cursor_ptr c = make_cursor(os);
+    serialize_begin(c);
+    while (!c->segments_done) {
+        serialize_segments(c);
+        co_await ss::maybe_yield();
+    }
+    serialize_replaced(c);
```

Comment:
If serialize() ran concurrently with housekeeping, this would risk outputting a contradictory manifest that listed the same segment in both the main list and in `replaced`.

We are perhaps relying on the property of ntp_archiver that it operates sequentially + will not do housekeeping at the same time as writing a manifest, but that is fragile.  The assumption will be violated if we refactor that to do background nagle'ed manifest updates, or if we are using this serialization method from the admin API.

Maybe we should wrap the whole thing in a mutex to make this simpler to reason about?  It would be almost always uncontended, since ntp_archiver is the only thing mutating the manifest.

Reply:
I added the protection against concurrent modification inside the `partition_manifest`. It will throw if manifest is updated concurrently during serialization process.

Reply:
I'm finding it really difficult to rule out scenarios like the one John mentioned. Because the correct behavior seem to all be predicated on what looks like _insync_offset being properly changed when something like `_segments` changes to enforce concurrency requirements it's concerning then to see methods like:

`partition_manifest::update`, that conditionally update `_segments` and `_insync_offset` where the input to this function is coming from some input stream.

unless we are in a super hot path, we should be erroring on the side of simplicity and debuggability, especially when it comes to issues of concurrency.